File: src/main/java/com/github/dkoval/leetcode/challenge/LongestPalindrome.java
package com.github.dkoval.leetcode.challenge;

/**
 * <a href="https://leetcode.com/problems/longest-palindrome/">Longest Palindrome</a>
 * <p>
 * Given a string s which consists of lowercase or uppercase letters, return the length of the longest
 * palindrome
 * that can be built with those letters.
 * <p>
 * Letters are case sensitive, for example, "Aa" is not considered a palindrome.
 * <p>
 * Constraints:
 * <ul>
 *  <li>1 <= s.length <= 2000</li>
 *  <li>s consists of lowercase and/or uppercase English letters only</li>
 * </ul>
 */
public interface LongestPalindrome {

    int longestPalindrome(String s);

    class LongestPalindromeRev2 implements LongestPalindrome {

        @Override
        public int longestPalindrome(String s) {
            int n = s.length();

            int[] counts = new int[52];
            for (int i = 0; i < n; i++) {
                int index = indexOf(s.charAt(i));
                counts[index]++;
            }

            int ans = 0;
            boolean tookOdd = false;
            for (int x : counts) {
                ans += x;
                if (x % 2 != 0) {
                    if (tookOdd) {
                        ans--;
                    }
                    tookOdd = true;
                }
            }
            return ans;
        }

        private int indexOf(char c) {
            // lowercase letters
            if (c >= 'a' && c <= 'z') {
                return c - 'a';
            }
            // uppercase letters
            return c - 'A' + 26;
        }
    }

    class LongestPalindromeRev3 implements LongestPalindrome {

        @Override
        public int longestPalindrome(String s) {
            int n = s.length();

            int[] counts = new int[52];
            for (int i = 0; i < n; i++) {
                int index = indexOf(s.charAt(i));
                counts[index]++;
            }

            int ans = 0;
            int odds = 0;
            for (int x : counts) {
                ans += x;
                if (x % 2 != 0) {
                    odds++;
                }
            }
            return ans - (odds > 1 ? odds - 1 : 0);
        }

        private int indexOf(char c) {
            if (c >= 'a' && c <= 'z') {
                return c - 'a';
            }
            return c - 'A' + 26;
        }
    }
}

File: src/main/kotlin/com/github/dkoval/leetcode/challenge/LongestPalindrome.kt
package com.github.dkoval.leetcode.challenge

object LongestPalindromeRev1 : LongestPalindrome {

    // O(N) time | O(ALPHA) space, where ALPHA is the number of characters in the alphabet
    override fun longestPalindrome(s: String): Int {
        val counts = mutableMapOf<Char, Int>()
        for (c in s) {
            counts[c] = counts.getOrDefault(c, 0) + 1
        }

        // Example: "xbcbxxcxcx"
        // b -> 2
        // c -> 3 (odd)
        // candidate: "bcccb"
        // x -> 5 (odd) - can form a better palindrome "bcxxxxxcb"
        // take 2 "c" (instead of 3) and include 5 "x"
        // also note that it doesn't matter in which order we discover "c" and "x", in both cases we decrement answer by 1
        var ans = 0
        var oddCountFound = false
        for (count in counts.values) {
            ans += count
            if (count % 2 != 0) {
                ans--
                oddCountFound = true
            }
        }
        return if (oddCountFound) ans + 1 else ans
    }
}